src/dstore/pmix_esh.c

```diff
@@ -0,0 +1,1374 @@
+static int _update_ns_elem(ns_track_elem_t *ns_elem, ns_seg_info_t *info);
+static int _put_ns_info_to_initial_segment(const char *nspace, pmix_sm_seg_t *metaseg, pmix_sm_seg_t *dataseg);
+static ns_seg_info_t *_get_ns_info_from_initial_segment(const char *nspace);
+static ns_track_elem_t *_get_track_elem_for_namespace(const char *nspace);
+static rank_meta_info *_get_rank_meta_info(int rank, seg_desc_t *segdesc);
+static uint8_t *_get_data_region_by_offset(seg_desc_t *segdesc, size_t offset);
+static void _update_initial_segment_info(void);
+static void _set_constants_from_env(void);
+static void _delete_sm_desc(seg_desc_t *desc);
+static int _pmix_getpagesize(void);
+static inline uint32_t _get_univ_size(const char *nspace);
+
+seg_desc_t *global_sm_seg_first, *global_sm_seg_last;
+static pmix_list_t namespace_info_list;
+static char *tmp_dir = NULL;
+static size_t initial_segment_size = 0;
+static size_t max_ns_num;
+static size_t meta_segment_size = 0;
+static size_t max_meta_elems;
+static size_t data_segment_size = 0;
+static int lockfd;
+static char *lockfile_name;
+
+/* If direct_mode is set, it means that we use linear search
+ * along the array of rank meta info objects inside a meta segment
+ * to find the requested rank. Otherwise,  we do a fast lookup
+ * based on rank and directly compute offset.
+ * This mode is called direct because it's effectively used in
+ * sparse communication patterns when direct modex is usually used.
+ */
+static int direct_mode = 0;
+
+static void ncon(ns_track_elem_t *p) {
+    p->meta_seg = NULL;
+    p->data_seg = NULL;
+    p->num_meta_seg = 0;
+    p->num_data_seg = 0;
+}
+
+static void ndes(ns_track_elem_t *p) {
+    _delete_sm_desc(p->meta_seg);
+    _delete_sm_desc(p->data_seg);
+}
+
+PMIX_CLASS_INSTANCE(ns_track_elem_t,
+                    pmix_list_item_t,
+                    ncon, ndes);
+
+static inline int _is_server(void)
+{
+    return (pmix_globals.server);
+}
+
+
+int _esh_init(void)
+{
+    int rc;
+    seg_desc_t *seg;
+
+    PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                         "%s:%d:%s", __FILE__, __LINE__, __func__));
+
+    rc = pmix_sm_init();
+    if (PMIX_SUCCESS != rc) {
+        PMIX_ERROR_LOG(rc);
+        return rc;
+    }
+
+    PMIX_CONSTRUCT(&namespace_info_list, pmix_list_t);
+    global_sm_seg_first = NULL;
+    global_sm_seg_last = NULL;
+    tmp_dir = strdup(pmix_tmp_directory());
+    _set_constants_from_env();
+    max_ns_num = (initial_segment_size - sizeof(size_t) - sizeof(int))/sizeof(ns_seg_info_t);
+    max_meta_elems = (meta_segment_size - sizeof(size_t))/sizeof(rank_meta_info);
+
+    /* create a lock file to prevent clients from reading while server is writing to the shared memory.
+     * This situation is quite often, especially in case of direct modex when clients might ask for data
+     * simultaneously.*/
+    lockfile_name = malloc(256);
+    snprintf(lockfile_name, 256, "%s/dstore_sm.lock", tmp_dir);
```

Comment:
@igor-ivanov 
We need to use jobid in the lock file name. It is possible that 2 simultaneous jobs will be running on the same node.